Implement forward AD for slogdet-related operations and fix their complex derivative

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #80074
* #80073
* #79838
* __->__ #79743
* #79742
* #79487
* #79486

The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

lezcano committed:
Implemnet forward AD for slogdet-related operations and fix their complex derivative

The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for slogdet-related operations and fix their complex derivative"


The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for slogdet-related operations and fix their complex derivative"


The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for slogdet-related operations and fix their complex derivative"


The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for slogdet-related operations and fix their complex derivative"


The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for slogdet-related operations and fix their complex derivative"


The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for slogdet-related operations and fix their complex derivative"


The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for slogdet-related operations and fix their complex derivative"


The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -229,7 +229,10 @@ Tensor lu_det_P(const Tensor& pivots) {
-  at::linalg_lu_factor_ex_out(const_cast<Tensor&>(LU), const_cast<Tensor&>(pivots), const_cast<Tensor&>(info), A);
+  // Optimisation: lu_factor_ex requires the input to be F-contig, otherwise it copies
+  // Use the transpose of if A is contiguous since det(A^T) = det(A)
+  // We limit this to real matrices, but it could also be implemented for complex matrices
```
We does it not work for complex inputs?

Lezcano replied:
We are saving the LU decomposition for the backward. In the backward I use `linalg.lu_solve`, which just has the `adjoint` flag. It could be done by exposing the `transpose` flag somehow but yeah.
Another way of going about it would be by marking this function as a fallthrough for complex conj, and checking that bit, but yeah, for the next time.

Lezcano replied:
Note that this is an optimisation that then we need to take into account in the backward. This same optimisation is done in `linalg_solve`.

nikitaved commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -257,7 +260,10 @@ Tensor det(const Tensor& self) {
+  // Optimisation: lu_factor_ex requires the input to be F-contig, otherwise it copies
+  // Use the transpose of if A is contiguous since det(A^T) = det(A)
+  // We limit this to real matrices, but it could also be implemented for complex matrices
+  at::linalg_lu_factor_ex_out(const_cast<Tensor&>(LU), const_cast<Tensor&>(pivots), const_cast<Tensor&>(info), A.is_contiguous() && !A.is_complex() ? A.mH() : A);
```
Having just `A.mT()` is `A.is_contiguous()` is sufficient, no?

Lezcano replied:
Not really as https://github.com/pytorch/pytorch/pull/79743#discussion_r901621302

nikitaved replied:
`A.conj().mT()`?

nikitaved reviewed (approved):
LGTM. Thank you, Mario!

lezcano committed:
Update on "Implement forward AD for slogdet-related operations and fix their complex derivative"


The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for slogdet-related operations and fix their complex derivative"


The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

[ghstack-poisoned]

lezcano committed:
Update on "Implement forward AD for slogdet-related operations and fix their complex derivative"


The derivative for complex inputs of slogdet was incorrect as it said
that the output `sign` was not differentiable. Same happened with
`logdet` which didn't even support complex inputs. Now these derivatives
are all correct and all the gradcheck tests pass.

We also add forward AD formulas for these functions.

We take this chance to announce in the docs that the derivatives of
these functions are much more better behaved than those from
`linalg.det`, so one should probably always use these functions over
`linalg.det` unless they have very good reasons not to do so.

[ghstack-poisoned]
